src/stash/build.rs

```diff
@@ -0,0 +1,207 @@
+#[derive(Debug, Clone, Deserialize, Serialize)]
+struct ProtoHash {
+    name: String,
+    md5: String,
+}
+
+const PROTO_DIRECTORY: &str = "protos";
+const PROTO_HASHES: &str = "protos/hashes.json";
+
+fn main() -> anyhow::Result<()> {
+    env::set_var("PROTOC", protobuf_src::protoc());
+
+    // Read in the persisted hashes from disk.
+    let hashes = fs::File::open(PROTO_HASHES).context("opening proto hashes")?;
+    let reader = BufReader::new(&hashes);
+    let hashes: Vec<ProtoHash> = serde_json::from_reader(reader)?;
+    let mut persisted: BTreeMap<String, String> =
+        hashes.into_iter().map(|e| (e.name, e.md5)).collect();
+
+    // Discover all of the protobuf files on disk.
+    let protos: BTreeMap<String, String> = fs::read_dir(PROTO_DIRECTORY)?
+        // If we fail to read one file, fail everything.
+        .collect::<Result<Vec<_>, _>>()?
+        .into_iter()
+        // Filter to only files with the .proto extension.
+        .filter(|entry| {
+            entry
+                .path()
+                .extension()
+                .map(|e| e.to_string_lossy().contains("proto"))
+                .unwrap_or(false)
+        })
+        .map(|file| {
+            let path = file.path();
+
+            // Hash the entire file.
+            let mut hasher = Md5::new();
+            let buffer = std::fs::read(&path).expect("To be able to read file");
+            hasher.update(buffer);
+
+            let name = path
+                .file_name()
+                .expect("To have file name")
+                .to_str()
+                .expect("UTF-8")
+                .to_string();
+            let hash = format!("{:x}", hasher.finalize());
+
+            (name, hash)
+        })
+        .collect();
+
+    // After validating our hashes we'll re-write the file.
+    let mut to_persist: Vec<ProtoHash> = Vec::new();
+
+    // Check the persisted hashes against what we just read in from disk.
+    for (name, hash) in protos {
+        match persisted.remove(&name) {
+            // Hashes have changed!
+            Some(og_hash) if hash != og_hash => {
+                anyhow::bail!(error_message(og_hash, hash, name));
+            }
+            // Found a proto file on disk that we didn't have persisted, we'll just persist it.
+            None => to_persist.push(ProtoHash { name, md5: hash }),
+            // We match!
+            Some(_) => to_persist.push(ProtoHash { name, md5: hash }),
+        }
+    }
+
+    // Check if there are any proto files we should have had hashes for, but didn't exist.
+    if !persisted.is_empty() {
+        anyhow::bail!("Have persisted hashes, but no files on disk? {persisted:#?}");
+    }
+
+    // Write the hashes back out to disk.
+    let file = fs::File::options()
+        .write(true)
+        .truncate(true)
+        .open(PROTO_HASHES)
+        .context("opening hashes file to write")?;
+    serde_json::to_writer(file, &to_persist).context("persisting hashes")?;
```

Comment:
```suggestion
    serde_json::to_writer_pretty(file, &to_persist).context("persisting hashes")?;
```

We need it to split on lines so a single migration doesn't rewrite the entire file.

Reply:
Love it, I didn't know `to_writer_pretty(...)` existed!